### ISSUE TITLE
Add host benchmarking

### ISSUE THREAD

**Issue by chris124567** [SiaFoundation/explored#122]

**Comment by CtrlAltDefeat94** [comment]
Not sure if it's planned already. But it would be helpful to be able to benchmark hosts while reusing the existing contract, unlike benchyd where each benchmark creates a new contract.

**Comment by chris124567** [comment]
From a cost perspective I think it makes sense to have that as the goal.

**Comment by n8maninger** [comment]
I disagree. Stick with simplicity and do a short contract and keep it 1 benchmark = 1 contract. Cost difference is negligible, but management is significantly simpler when you can assume a new contract will need to be formed.

**Comment by CtrlAltDefeat94** [comment]
@n8maninger could you elaborate how the cost is negligible to create a new contract every time?

I did 2 benchmarks, one with 1 sector and one with 10 sectors. It looks like the the cost difference isn't too much. Both around 0.27SC
https://siascan.com/contract/bd469bdc5e7c40caabcd350845d5417f71749de661454f8ec7330554b67e3683
https://siascan.com/contract/4c60765acba8d85738feb56d3d28433af428444a59a37250015a1b93269e80e6

Running 10 benchmarks with each their own contract would cost around 2.7SC. If a single contract 10x the size would barely make a difference compared to a single contract, how would reusing contracts not make a significant difference?

Another thing. I feel like I would be spamming the blockchain a lot if I wouldn't reuse contracts. If I were to benchmark all hosts for example 4 times a day, with 500 hosts I would be creating 60K contracts a month, with roughly 4K of them active at a time. That's more than 10% of the current active contracts on the entire network.

I'm still a newbie on this side, any feedback/elaboration is welcome :)

**Comment by n8maninger** [comment]
Contract ID | Sectors | Initial Allowance | Final Allowance | Renter Cost | Cost Per Sector
-- | -- | -- | -- | -- | --
bd469bdc5e7c40caabcd350845d5417f71749de661454f8ec7330554b67e3683 | 1 | 0.002173 | 0.0007788 | 0.001394 | 0.001394
4c60765acba8d85738feb56d3d28433af428444a59a37250015a1b93269e80e6 | 10 | 0.02173 | 0.007788 | 0.01394 | 0.001394

Looking at both of those contracts, the second one has about 10x more spent than the first. This doesn't include the Siafund tax, but I would still consider it a negligible difference.

Spamming the blockchain isn't too terribly concerning. Onboarding more renters means more active contracts too, so I generally consider that a good thing. 

Having said that, I would consider the goals when trying to benchmarking the whole network 4x a day, though. Providing trends over a longer period of time will likely provide just as useful data than frequent benchmarks over such a short period. Also significantly increases infrastructure requirements as the network grows. It's largely fine at 500 hosts, but what about 5000?

**Comment by CtrlAltDefeat94** [comment]
As described in my grant request, I want to build a system to check hosts on interval and warn them if there are any issues.
I want to use benchmarks to determine whether hosts are actually writable. Since having free space and being online doesn't always confirm a host is able to receive data.
I'm still not sure about the interval to use, 4x was an example.

I might stick to only running benchmarks on hosts that subscribed to alerts.
Maybe there's an entirely different approach that would work better. I'm open to suggestions, but that's probably something for another place.
